src/crimson/osd/ops_executer.h

```diff
@@ -111,15 +140,17 @@ class OpsExecuter {
   // when operation requires this division, some variant of `with_effect()`
   // should be used.
   struct effect_t {
-    virtual osd_op_errorator::future<> execute() = 0;
+    // an effect can affect PG, i.e. create a watch timeout
+    virtual osd_op_errorator::future<> execute(Ref<PG> pg) = 0;
     virtual ~effect_t() = default;
   };
 
   ObjectContextRef obc;
   const OpInfo& op_info;
   const pg_pool_t& pool_info;  // for the sake of the ObjClass API
   PGBackend& backend;
-  const MOSDOp& msg;
+  ceph::static_ptr<ExecutableMessage,
```

Comment:
Why aren't we using the ref-counted messages here?

Reply:
We need to decouple from `MOSDOp` but without altering its inheritance hierarchy.